### PR TITLE
[FixCode] Add a diagnosis/fixit to help users convert void pointers.

### PR DESCRIPTION
**Pull request by nkcsgexi**
This patch warns the usage of Unsafe[Mutable]Pointer<Void>, and fix users' code by replacing it with Unsafe[Mutable]RawPointer.

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…Pointer<Void> to Unsafe[Mutable]RawPointer. rdar://27730408

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci smoke test OS X platform

**Comment by nkcsgexi** [comment]
@jrose-apple mind reviewing this?

**Comment by nkcsgexi** [comment]
@swift-ci smoke test OS X platform

**Comment by nkcsgexi** [comment]
@swift-ci smoke test OS X platform

File: lib/Sema/TypeCheckType.cpp [hunk 1]
```diff
@@ -408,6 +408,25 @@ Type TypeChecker::resolveTypeInContext(
...
+  auto *BGT = Ty->getAs<BoundGenericType>();
+  if (!BGT)
+    return false;
+  if (BGT->getDecl() != Ctx.getUnsafePointerDecl() &&
+      BGT->getDecl() != Ctx.getUnsafeMutablePointerDecl())
+    return false;
+  IsMutable = BGT->getDecl() == Ctx.getUnsafeMutablePointerDecl();
+  assert(BGT->getGenericArgs().size() == 1);
+  auto CanTy = BGT->getGenericArgs().front()->getCanonicalType();
+  return CanTy->isVoid();
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckType.cpp]: I think you can skip the `getCanonicalType` step here.

**Comment by jrose-apple** [comment]
I'm a little unhappy with hardcoding this in the compiler, and I'm not sure that `UnsafeMutablePointer<Void>` is *never* correct, but the implementation looks fine. @atrick, any comments?

**Comment by atrick** [comment]
Hard-coding this into the compiler is a last ditch effort to get decent diagnostics (I spent lots of time last week exploring the limits of availability annotations).

I'm not sure "deprecated" is the right language. UnsafePointer<Void> will always be a valid type in generic contexts, but people should be strongly discouraged from using it in a concrete bound generic. If they do, it will cause a number of confusing downstream errors, and the diagnostics for those errors will be better if they first rename to UnsafeRawPointer.

**Comment by atrick** [comment]
@jrose-apple `UnsafePointer<Void>` is a valid type (not really deprecated), but if a user is spelling that type, then they really should be using `UnsafeRawPointer`. I would be fine with a warning:
"Unsafe%0Pointer<Void> has been replaced by Unsafe%0RawPointer"
We do need a warning to help with code migration because the `void*` import type has changed and `UnsafePointer` conversion is restricted.

**Comment by nkcsgexi** [comment]
@atrick Should the warning be "Unsafe%0Pointer has been replaced by Unsafe%0RawPointer" or "Unsafe%0Pointer<Void> has been replaced by Unsafe%0RawPointer"? I thought the later is the case we are handling here specifically.

**Comment by jrose-apple** [comment]
:-) I think you both fell prey to Markdown's HTML embedding. "Unsafe%0Pointer\<Void\>" (written `Unsafe%0Pointer\<Void\>`).

**Comment by atrick** [comment]
I think we both didn't quote the warning string properly. I think it should be
`Unsafe%0Pointer<Void> has been replaced by Unsafe%0RawPointer`

https://github.com/apple/swift/pull/4114 has been merged.

**Comment by nkcsgexi** [comment]
@jrose-apple @atrick Right, I realized what happened the moment I saw my message posting.

**Comment by nkcsgexi** [comment]
@swift-ci test

**Comment by nkcsgexi** [comment]
@swift-ci test

**Comment by nkcsgexi** [comment]
@swift-ci test

**Comment by nkcsgexi** [comment]
Slave went offline during the build.
@swift-ci  test

**Pull request merged**